src/python/pants/pantsd/pailgun_server.py

```diff
@@ -178,19 +196,74 @@ def handle_request(self):
     with self.lifecycle_lock():
       self._handle_request_noblock()
 
+  def _should_poll_forever(self, timeout):
+    return timeout < 0
+
+  def _should_keep_polling(self, timeout, time_polled):
+    return self._should_poll_forever(timeout) or time_polled < timeout
+
+  @contextmanager
+  def ensure_request_is_exclusive(self, environment, request):
+    """
+    Ensure that this is the only pants running.
+
+    We currently don't allow parallel pants runs, so this function blocks a request thread until
+    there are no more requests being handled.
+    """
+    # TODO add `did_poll` to pantsd metrics
+
+    timeout = float(environment.pop('PANTSD_REQUEST_TIMEOUT_LIMIT'))
+
+    @contextmanager
+    def yield_and_release(time_waited):
+      try:
+        self.logger.debug("request lock acquired {}".format("on the first try" if time_waited == 0 else "in {}s".format(time_waited)))
+        yield
+      finally:
+        self.handling_request_lock.release()
+        self.logger.debug("released request lock")
+
+    time_polled = 0.0
+    polling_interval = 1.0 # check the lock every second
+    self.logger.debug("request {} is trying to aquire the request lock".format(request))
+
+    # Optimistically try to acquire the lock without blocking, in case we are the only request being handled.
+    # This could be merged into the while below, but separating this special case for logging helps.
```

Comment:
Escape `while` with backticks to make clear that it's a loop and not the English word while.

Reply:
Good catch!